geowebcache/mbtiles/src/main/java/org/geowebcache/mbtiles/layer/MBTilesLayer.java

```diff
@@ -0,0 +1,349 @@
+        try {
+            switch (metadataFormat) {
+                case PNG:
+                    mimeType = MimeType.createFromFormat("image/png");
+                    break;
+                case JPEG:
+                case JPG:
+                    mimeType = MimeType.createFromFormat("image/jpeg");
+                    break;
+                case PBF:
+                    mimeType = ApplicationMime.mapboxVector;
+                    vectorTiles = true;
+                    break;
+            }
+        } catch (MimeException e) {
+            throw new RuntimeException(e);
+        }
+        if (mimeType != null) {
+            return Collections.singletonList(mimeType);
+        }
+
+        return Collections.emptyList();
+    }
+
+    private HashMap<String, GridSubset> createGridSubsets(final GridSetBroker gridSetBroker) {
+        GridSet gridSet;
+
+        DefaultGridsets defaultGridSet = new DefaultGridsets(true, true);
+
+        if (tileSize <= 0) {
+            tileSize = DEFAULT_TILE_SIZE;
+        }
+
+        switch (tileSize) {
+            case TILE_SIZE_256:
+                gridSet = defaultGridSet.worldEpsg3857();
+                break;
+            case TILE_SIZE_512:
+                gridSet = defaultGridSet.worldEpsg3857x2();
+                break;
+            default:
+                throw new IllegalArgumentException("Unsupported tileSize: " + tileSize);
+        }
+        Integer minZoom = cacheInfo.getMinZoom();
+        Integer maxZoom = cacheInfo.getMaxZoom();
+
+        GridSubset subSet =
+                GridSubsetFactory.createGridSubSet(gridSet, this.layerBounds, minZoom, maxZoom);
+
+        HashMap<String, GridSubset> subsets = new HashMap<>();
+        subsets.put(gridSet.getName(), subSet);
+        return subsets;
+    }
+
+    /** @see org.geowebcache.layer.TileLayer#getTile(org.geowebcache.conveyor.ConveyorTile) */
+    @Override
+    public ConveyorTile getTile(final ConveyorTile tile)
+            throws IOException, OutsideCoverageException {
+
+        long[] tileIndex = tile.getTileIndex();
+        int zl = (int) tileIndex[2];
+        int row = (int) tileIndex[1];
+        int column = (int) tileIndex[0];
+        MBTilesTile loadedTile = mbTilesFile.loadTile(zl, column, row);
+        byte[] content = loadedTile.getData();
+        if (content != null) {
+            if (cacheInfo.getFormat() == MBTilesMetadata.t_format.PBF) {
+                content = getPbfFromTile(content);
+            }
+
+            tile.setBlob(new ByteArrayResource(content));
+            tile.setCacheResult(CacheResult.HIT);
+        } else {
+            tile.setCacheResult(CacheResult.MISS);
+            throw new OutsideCoverageException(tile.getTileIndex(), 0, 0);
+        }
+
+        saveExpirationInformation((int) (tile.getExpiresHeader() / 1000));
+
+        return tile;
+    }
+
+    private byte[] getPbfFromTile(byte[] raw) throws IOException {
+        if (!UNZIP_CONTENT) {
```

Comment:
Not sure how well this plays againt the rest of GWC though, to make it zip preservation work, we'd have to tell the services that the payload is already gzip compressed, and then, decide what to do based on whether the HTTP client declares to support gzip compression or not.